### DISCUSSION TITLE
Failed assert when using app layer with OOT controller

### DISCUSSION THREAD

**Discussion by Arvli** [bitcraze/discussions#997] [Q&A]
I am writing a controller for the Crazyflie, and when running it I sometimes get the error:
[crazyflie_server-4] [INFO] [1698127276.887339203] [cf1]: SYS: The system resumed after a failed assert [WARNING]
[crazyflie_server-4] [INFO] [1698127276.895276597] [cf1]: SYS: Assert failed at /home/arvid/Documents/exjobb/crazyflie-firmware/src/hal/src/radiolink.c:164

or:
[crazyflie_server-4] [INFO] [1698127215.971867005] [cf1]: SYS: The system resumed after a failed assert [WARNING]
[crazyflie_server-4] [INFO] [1698127215.978710486] [cf1]: SYS: Assert failed at /home/arvid/Documents/exjobb/crazyflie-firmware/vendor/FreeRTOS/timers.c:532

My belief has been that it has to do with my algorithm taking too much time, but when logging the CPU usage it seems to work. 
The problem is very inconsistent. A couple of days ago I could fly for several minutes without problem, but today I could hardly start the crazyflie before it rebooted. 

I have seen the similar post on the forum 
https://forum.bitcraze.io/viewtopic.php?t=4797 
but I have outsourced the heavy control part to the app task, and set its priority to 2. So I think all timecritical processes should be able to interrupt it? And I do not sending any data from my app, more than the standard logging. 
I am also using quite a lot of memory, but I have allocated enough stack according to the task dump. 
Any ideas on what could be the problem? Or what tests I can do to get more information about it?

**Comment by krichardsson** [comment]
It is hard to give any good advice.

> My belief has been that it has to do with my algorithm taking too much time

Yes, the assert in radiolink.c often indicate that there is a lack of CPU time. The problem might be that something is taking too much CPU time, but it is perhaps also possible that something is locked/blocked due to a full queue or similar. For example, if you are pushing packets to the radio you might want to use the non blocking flavor (`appchannelSendDataPacket()` as opposed to `appchannelSendDataPacketBlock()`)

> The problem is very inconsistent.

Maybe you can strip off some parts to figure out what is causing the problems? 

> I have outsourced the heavy control part to the app task, and set its priority to 2. So I think all timecritical processes should be able to interrupt it?

This sounds like a good approach.

We have seen this type of problems from time to time and we fixed an issue (in PR https://github.com/bitcraze/crazyflie-firmware/pull/1308) that is related to the lighthouse system and inefficient use of the UART. If you are using the lighthouse system, 4 base stations and the PR is not part of the code base, this might be what you see. It could explain the inconsistency as it depends on how many base stations that are visible.
